### PR TITLE
V4 migration and unmigration tool

### PR DESCRIPTION
**Pull request by arc-vmware**
Adding the tool for migration from v4 to categorized blockchain and vice versa

### CODE REVIEW

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (cloudnoize) [inline kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp]: if the genesis block is greater than one, we'll need to perform a special pruning handling.

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (cloudnoize) [inline kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp:150]: if we return here, if I understand the code, it will stop the storage thread.
Does it mean that the migration is over?

Comment (arc-vmware) [reply kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp:150]: Yes this means the migration is over. the logic is updated to take all the cases.

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (arc-vmware) [reply kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp]: Thanks @cloudnoize for this comment. As discussed we will wait on the final decision and then add pruning handling.

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (cloudnoize) [inline kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp:50]: do we want to assert that input-rocksdb-path != output-rocksdb-path ?

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (cloudnoize) [inline kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp:153]: the previous check for termination and this predicate satisfy if ```max_write_batch_id_== max_batch_id``` 
and since updating max_read_batch_id_ and read_completed_ is not atomic, can we have a race where max_write_batch_id_ == max_batch_id and read_completed_ is false though in practice is about to be true yet this thread performs those checks before the other sets it to true?

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (arc-vmware) [reply kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp:50]: @cloudnoize , thanks for the comment. I am not checking this, since the DB can be the same aswell. So input DB path and output DB path can be same.

File: kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp
Comment (arc-vmware) [reply kvbc/tools/migrations/v4migration_tool/src/migration_bookeeper.cpp:153]: Thanks @cloudnoize , this is a good catch. The termination condition should not have equal to in it.

**Review by cloudnoize — APPROVED** [review]
LGTM!!

**Pull request merged**
